core/arch/riscv/mm/core_mmu.c

```diff
@@ -0,0 +1,639 @@
+		core_mmu_table_write_barrier();
+		satp |= ((uint64_t)map->asid << RISCV_SATP_ASID_SHIFT);
+		write_satp(satp);
+	} else {
+		core_mmu_entry_set(pte, 0);
+		core_mmu_table_write_barrier();
+	}
+
+	thread_unmask_exceptions(exceptions);
+}
+
+void core_mmu_get_user_va_range(vaddr_t *base, size_t *size)
+{
+	assert(user_va_idx != -1);
+
+	if (base)
+		*base =  (vaddr_t)1 << CORE_MMU_USER_TABLE_SHIFT;
+
+	if (size)
+		*size = (RISCV_PTES_PER_PT - 1) << CORE_MMU_USER_TABLE_SHIFT;
+}
+
+void core_mmu_get_user_pgdir(struct core_mmu_table_info *pgd_info)
+{
+	vaddr_t va_range_base;
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	pgt_t *pgt = core_mmu_get_ta_pgt_va(prtn);
+
+	core_mmu_get_user_va_range(&va_range_base, NULL);
+	/*
+	 * TODO: sync with core_mmu_set_info_table
+	 *... dir_info->level + 1 ... in core_mmu.c
+	 */
+	core_mmu_set_info_table(pgd_info, 0, va_range_base, pgt);
+}
+
+void core_mmu_create_user_map(struct user_mode_ctx *uctx,
+			      struct core_mmu_user_map *map)
+{
+	struct core_mmu_table_info tbl_info;
+
+	core_mmu_get_user_pgdir(&tbl_info);
+	memset(tbl_info.table, 0, RISCV_MMU_PGT_SIZE);
+	core_mmu_populate_user_map(&tbl_info, uctx);
+	map->user_map = virt_to_phys(tbl_info.table);
+	map->asid = uctx->vm_info.asid;
+}
+
+void core_mmu_get_user_map(struct core_mmu_user_map *map)
+{
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	pte_t *pte = core_mmu_get_user_mapping_entry(prtn);
+
+	map->user_map = core_mmu_pte_pa(pte);
+
+	if (map->user_map)
+		map->asid = (read_satp() >> RISCV_SATP_ASID_SHIFT) &
+			     RISCV_SATP_ASID_MASK;
+	else
+		map->asid = 0;
+}
+
+bool core_mmu_user_mapping_is_active(void)
+{
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+	bool ret = false;
+	pte_t *pte;
+	uint32_t exceptions;
+
+	exceptions = thread_mask_exceptions(THREAD_EXCP_ALL);
+	pte = core_mmu_get_user_mapping_entry(prtn);
+	ret = core_mmu_entry_is_valid(pte);
+	thread_unmask_exceptions(exceptions);
+
+	return ret;
+}
+
+void core_init_mmu_prtn(struct mmu_partition *prtn, struct tee_mmap_region *mm)
+{
+	core_init_mmu_prtn_tee(prtn, mm);
+	core_init_mmu_prtn_ta(prtn);
+}
+
+void core_init_mmu(struct tee_mmap_region *mm)
+{
+	uint64_t max_va = 0;
+	size_t n;
+
+	COMPILE_TIME_ASSERT(CORE_MMU_BASE_TABLE_SHIFT ==
+			    CORE_MMU_SHIFT_OF_LEVEL(CORE_MMU_BASE_TABLE_LEVEL));
+	COMPILE_TIME_ASSERT((RISCV_MMU_MAX_PGTS * RISCV_MMU_PGT_SIZE) ==
+			    sizeof(xlat_tables));
+
+	/* Initialize default pagetables */
+	core_init_mmu_prtn_tee(&default_partition, mm);
+
+	for (n = 0; !core_mmap_is_end_of_table(mm + n); n++) {
+		vaddr_t va_end = mm[n].va + mm[n].size - 1;
+
+		if (va_end > max_va)
+			max_va = va_end;
+	}
+
+	set_user_va_idx(&default_partition);
+
+	core_init_mmu_prtn_ta(&default_partition);
+
+	assert(max_va < BIT64(RISCV_MMU_VA_WIDTH));
+}
+
+void core_init_mmu_regs(struct core_mmu_config *cfg)
+{
+	struct mmu_partition *prtn = core_mmu_get_prtn();
+
+	cfg->satp = core_mmu_pgt_to_satp(prtn->asid,
```

Comment:
`core_init_mmu_regs()` is rather short, perhaps `core_mmu_pgt_to_satp()` could be open coded here instead. It might make it easier to follow what's going on. No big deal, keep it as it is if you prefer that.